### DISCUSSION TITLE
Failing to Integrate the Drv::LinuxI2cDriver Component Correctly (General Help Needed with I2C in F')

### DISCUSSION THREAD

**Discussion by John-P-Cargille** [nasa/fprime#2164] [Q&A]
Hey All,

Our team has been working on flight software for a couple weeks now; last week I was trying to define and instantiate an active component which would hook up with the passive LinuxI2cDriver component to send a basic I2C read command to our physical ADCS motherboard (MB). I successfully built a project with my active "AdcsComponent" component defined, and then created a deployment in which I instantiated an "adcsComponent" and a "linuxI2cDriver" component. I then went through the cross-compilation steps as laid out by the led-blinker workshop, copying an aarch64-linux version of the deployment onto a Raspberry Pi 4 Model B, without any errors.

After cross-compilation, I learned how to set up the ADCS MB for testing; I won't go into our set-up in depth unless requested, but we were going off of pin connections listed in the manual for said ADCS MB, the set-up had worked in the past for people executing I2C commands through a computer software and driver configuration, and I had multiple people regularly checking my circuitry throughout the process. I was able to run the fprime-gds as described in the led-blinker workshop--with the Pi and ADCS MB wired up and both powered at the time--and successfully sent the user-defined "adcsComponent.AdcsI2cRead" command through it; this returned a user-defined warning high event linked to the returned "I2cStatus" enum being a "I2C_OPEN_ERR" (the address used being the address of the register within the ADCS MB that holds its firmware version).

<img width="945" alt="Screenshot 2023-07-28 154424" src="https://github.com/nasa/fprime/assets/116765885/d6ded3b0-92e7-4aa6-8752-afa682cf9047">

Using a multimeter, an oscilloscope, and the "i2c-tools" debian package, my peers and I were able to determine the following:
1) The circuit is powered in the way we expect it to be
2) The Pi is in fact generating I2C clock and I2C data activity on the correct connections when we employ the i2c-tools command "i2cdetect" (this does not occur when the read command is sent through the F' GDS)
3) The Pi is aware of the device at the address of the ADCS MB, as we have confirmed using its manual

Let me know if you have any further suggestions for testing. All this considered, it would seem that the problem rests not in our testing set-up, but in the code I have written.

In light of this, I first made sure to check that the "I2C_OPEN_ERR" was not the default value for I2cStatus, which it is not. I also looked into whether I should send the I2C bus address or the peripheral device's register address as a command parameter off of the advice of a peer, but it didn't seem to make a difference what address I actually submitted. In looking more closely at the LinuxI2cDriver component implementation, I found that, in the .cpp file, the handler for the "read" port doesn't actually do anything before checking to see if the device is open, but that leaves me at a loss as to how and where to open the device manually. The same .cpp file defines an "open" method but it's difficult for me to parse exactly what this does, and I wasn't able to successfully employ it in further testing of the MB.

So, my question is two-fold: first, looking at the project files supplied, do any immediate problems jump out? Secondly, are there any resources for developing FSW with I2C connections using F', either already created or in development? All I could find in my searching were some forum posts from two years ago--none of which seemed to be dealing with the same problem I was experiencing--a nonfunctional link to a community-made I2C tutorial, [this page of the F' User's Guide](https://nasa.github.io/fprime/UsersGuide/best/app-man-drv.html) which mentions using I2C, and the LinuxI2cDriver itself, which doesn't seem to have any additional documentation attached to it. I also did do some looking into the RPI demo, but encountered further difficulties and confusion trying to apply the logic used in that implementation code to my own.

Thanks to anyone who looks into it.

[build-fprime-automatic-native.zip](https://github.com/nasa/fprime/files/12217205/build-fprime-automatic-native.zip)
[Components.zip](https://github.com/nasa/fprime/files/12217178/Components.zip)
[CurrentDeployment.zip](https://github.com/nasa/fprime/files/12217179/CurrentDeployment.zip)
[CMakeLists.txt.zip](https://github.com/nasa/fprime/files/12217199/CMakeLists.txt.zip)
[project.cmake.zip](https://github.com/nasa/fprime/files/12217202/project.cmake.zip)

A Note on General Cross-Compilation: Someone might find it useful to know that cross-compiling following a previous version of the led-blinker workshop produced an error for a couple of us on the team in which duplicate cmake files were generated (seemingly because of the inclusion of the fprime-arm-linux.git submodule), but I no longer encountered this error when following the updated workshop "running-on-hardware" instructions; figured I'd mention that we were also experiencing this error in retrospect, both to add a data point to the issue and to make it clear that this fix was appreciated!

**Comment by thomas-bc** [comment]
Hey @John-P-Cargille, just a recommendation, would it be possible to share an SCM link (GitHub or otherwise) instead of zipping _some_ files? That would make it much _**much**_ easier to inspect on our side.
As for your last note, fprime-arm-linux used to be a submodule you had to add to your fprime project, but the toolchain files have since been [added to core fprime](https://github.com/nasa/fprime/pull/2101). This is why you had an error following old versions of the tutorials, CMake doesn't like when files are duplicated.

**Reply by John-P-Cargille** [reply]
Gotcha. Part of the difficulty is that we currently have our flight software in a private repository, with the thought being that the software we write may eventually reach a point at which revealing it would infringe on certain non-disclosure agreements we have with the manufacturers of our flight hardware. For now, the design is pretty basic, so I've gone ahead and worked around this by duplicating the private repository files on my local machine, removing the .git and .gitmodules directories, and pushing the project to [this public repository](https://github.com/John-P-Cargille/AdcsTestProject_8_1). I suppose in the future it will probably be most advisable to create minimal viable products which still replicate the issues we are experiencing when reaching out for public help? Let me know if you have any advice on the matter.
Thaaaat makes a lot of sense; thanks for letting me know what the issue was.

**Reply by SterlingPeet** [reply]
At GT, we use submodules to handle things that might be sensitive.  Usually the code that might be covered by an NDA gets wrapped into a one or two components that live in their own repository with a simple demo deployment that can be used for basic integration testing of those modules only.

The main flight deployment repositories will reference these as submodules, rather than keeping the sensitive code in the main repository.

Additionally, we typically set up a basic testing deployment in that repository that doesn't use the referenced submodules.  That lets us get all the platform kinks worked out with F Prime core, without dragging in sensitive code.
